File: pyeth/eth_data/eth_data/tx_processor/data_models/trace_models.py
from dataclasses import dataclass, field
from typing import List, Optional
from eth_typing import ChecksumAddress


@dataclass
class TraceCall:
    type: str
    from_address: str
    to_address: str
    value: Optional[int]
    gas: int
    gas_used: int
    input: str
    output: str
    error: Optional[str]
    calls: List['TraceCall'] = field(default_factory=list)


@dataclass
class TransactionTrace:
    trace: TraceCall


@dataclass
class InternalTransaction:
    from_address: ChecksumAddress
    to_address: Optional[ChecksumAddress]
    value: int
    gas: int
    gas_used: int
    depth: int
    trace_type: str
    call_type: Optional[str]
    error: Optional[str]
    

@dataclass
class TraceOperation:
    type: str
    from_address: ChecksumAddress
    to_address: ChecksumAddress
    value: int 
    gas: int
    gas_used: int
    input: str

File: reth/tx_processor/src/tx_processor/data_models/trace_models.rs
use alloy_primitives::{Address, U256};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InternalTransaction {
    pub from_address: Address,
    pub to_address: Option<Address>,
    pub value: U256,
    pub gas: u64,
    pub gas_used: u64,
    pub trace_type: String,
    pub call_type: Option<String>,
    pub depth: u32,
    pub error: Option<String>,
}
